crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs

```diff
@@ -0,0 +1,220 @@
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::internal_prelude::*;
+
+pub type SimpleTransferAnalyzer = SimpleTransferStateMachine;
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum SimpleTransferStateMachine {
+    #[default]
+    InitialState,
+    AccessControllerProofCreated,
+    FeeLockPerformed,
+    ResourcesWithdrawn,
+    ResourcesInBucket,
+    DepositPerformed,
+    InvalidState,
+}
+
+impl ManifestStaticAnalyzer for SimpleTransferStateMachine {
+    type Initializer = ();
+    type Output = ();
+    type PermissionState = Self;
+    type RequirementState = Self;
+
+    fn new(
+        _: Self::Initializer,
+    ) -> (Self, Self::PermissionState, Self::RequirementState) {
+        Default::default()
+    }
+
+    fn output(self) -> Self::Output {}
+
+    fn process_instruction(&mut self, context: AnalysisContext<'_>) {
+        self.transition(context.instruction());
+    }
+}
+
+impl ManifestAnalyzerRequirementState for SimpleTransferStateMachine {
+    fn requirement_state(&self) -> RequirementState {
+        match self {
+            SimpleTransferStateMachine::DepositPerformed => {
+                RequirementState::Fulfilled
+            }
+            SimpleTransferStateMachine::InitialState
+            | SimpleTransferStateMachine::AccessControllerProofCreated
+            | SimpleTransferStateMachine::FeeLockPerformed
+            | SimpleTransferStateMachine::ResourcesWithdrawn
+            | SimpleTransferStateMachine::ResourcesInBucket => {
+                RequirementState::CurrentlyUnfulfilled
+            }
+            SimpleTransferStateMachine::InvalidState => {
+                RequirementState::PermanentlyUnfulfilled
+            }
+        }
+    }
+
+    fn process_instruction(&mut self, context: AnalysisContext<'_>) {
+        self.transition(context.instruction());
+    }
+}
+
+impl ManifestAnalyzerPermissionState for SimpleTransferStateMachine {
+    fn all_instructions_permitted(&self) -> bool {
+        !matches!(self, Self::InvalidState)
+    }
+
+    fn process_instruction(&mut self, context: AnalysisContext<'_>) {
+        self.transition(context.instruction());
+    }
+}
+
+impl SimpleTransferStateMachine {
+    pub fn transition(&mut self, instruction: &GroupedInstruction) {
+        let next_state = match (&self, instruction) {
+            // Initial State -> Access Controller Proof State transition.
+            (
+                Self::InitialState,
+                GroupedInstruction::InvocationInstructions(
+                    InvocationInstructions::CallMethod(CallMethod {
+                        address: ManifestGlobalAddress::Static(address),
+                        method_name,
+                        ..
+                    }),
+                ),
+            ) if matches!(
+                address.as_node_id().entity_type().map(Into::into),
```

Comment:
MINOR: I'd consider adding an extension method to `pub(crate) trait GlobalAddressExt { grouped_entity_type(&self) -> Option<GroupedEntityType> }`

Reply:
I think instead of doing that, using typed invocations here would make a lot more sense and we no longer need to do the entity type check here since it's been done for us when converting into the typed invocations.